### PR TITLE
Remove no return warning from val/test step

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6137

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/overrides/base.py [hunk 1]
```diff
@@ -54,15 +54,15 @@ def forward(self, *inputs, **kwargs):
             # ddp_plugin ``post_training_step`` hook
             if not self.module.automatic_optimization:
                 self.module.trainer.model.require_backward_grad_sync = False
-            warn_if_output_is_none(output, "training_step")
+
+            if self.module.automatic_optimization:
+                warn_if_output_is_none(output, "training_step")
```
Comment (rohitgr7) [inline pytorch_lightning/overrides/base.py:59]: I wonder why do we have this warning here. Isn't this enough? https://github.com/PyTorchLightning/pytorch-lightning/blob/0456b4598f5f7eaebf626bca45d563562a15887b/pytorch_lightning/trainer/training_loop.py#L756-L757

File: tests/overrides/test_data_parallel.py [hunk 2]
```diff
@@ -57,18 +57,25 @@ def test_lightning_wrapper_module_warn_none_output(wrapper_class):
...
 
-    with pytest.warns(UserWarning, match="Your validation_step returned None"):
+    with pytest.warns(None, match="Your validation_step returned None") as record:
         pl_module.running_stage = RunningStage.EVALUATING
         wrapped_module()
+        assert not record
+
+    with pytest.warns(UserWarning, match="Your predict returned None"):
+        pl_module.running_stage = RunningStage.PREDICTING
+        wrapped_module()
```
Comment (rohitgr7) [inline tests/overrides/test_data_parallel.py:78]: I believe predict can't return `None` right?? I haven't tried it out yet.

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/base.py:59]: could you include this in the test below? automatic -> warning, manual -> no warning
thanks

**Review by awaelchli — APPROVED** [review]
It's strange, I wrote this code when refactoring and it didn't occur to me that the warning does not apply for eval steps. 
However, it looks like it was the behavior even before the refactor
Here for DDP: https://github.com/PyTorchLightning/pytorch-lightning/blob/748a74e261585bb571bfc6fa1cc85d33fb44f163/pytorch_lightning/overrides/data_parallel.py#L203

and here for DP: https://github.com/PyTorchLightning/pytorch-lightning/blob/748a74e261585bb571bfc6fa1cc85d33fb44f163/pytorch_lightning/overrides/data_parallel.py#L283

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/overrides/base.py:59]: yes doing that. but my question here is if we already have a similar warning inside training_loop, then why do we need it in overrides/base.py?

File: tests/overrides/test_data_parallel.py (hunk 2 above)
Comment (rohitgr7) [reply tests/overrides/test_data_parallel.py:78]: also, there is no warning when predict doesn't return anything in non-distributed env. should I add that?

Comment (awaelchli) [reply tests/overrides/test_data_parallel.py:78]: it should be fine not to warn because in contrast to the training step there are no undesired side effects afaik. The reason for warning in the case of training_step is so that the user is aware they are actually skipping steps and not do optimization

File: tests/overrides/test_data_parallel.py [hunk 3]
```diff
@@ -57,18 +57,25 @@ def test_lightning_wrapper_module_warn_none_output(wrapper_class):
...
     pl_module.validation_step.return_value = None
     pl_module.test_step.return_value = None
+    pl_module.predict.return_value = None
 
     with pytest.warns(UserWarning, match="Your training_step returned None"):
         pl_module.running_stage = RunningStage.TRAINING
         wrapped_module()
 
-    with pytest.warns(UserWarning, match="Your test_step returned None"):
+    with pytest.warns(None, match="Your test_step returned None") as record:
```
Comment (Borda) [inline tests/overrides/test_data_parallel.py:66]: is this kind of reverse check that the warning is not raised?

File: tests/overrides/test_data_parallel.py (hunk 2 above)
Comment (rohitgr7) [reply tests/overrides/test_data_parallel.py:78]: hmm. I see it isn't doing anything special.. just appending the predictions... no need :+1:

File: tests/overrides/test_data_parallel.py (hunk 3 above)
Comment (rohitgr7) [reply tests/overrides/test_data_parallel.py:66]: `assert not record` is the actual the reverse check

File: tests/overrides/test_data_parallel.py [hunk 4]
```diff
@@ -57,18 +57,25 @@ def test_lightning_wrapper_module_warn_none_output(wrapper_class):
...
         wrapped_module()
+        assert not record
 
-    with pytest.warns(UserWarning, match="Your validation_step returned None"):
+    with pytest.warns(None, match="Your validation_step returned None") as record:
         pl_module.running_stage = RunningStage.EVALUATING
         wrapped_module()
+        assert not record
+
+    with pytest.warns(UserWarning, match="Your predict returned None"):
```
Comment (carmocca) [inline tests/overrides/test_data_parallel.py:76]: This should not warn, right?

Comment (rohitgr7) [reply tests/overrides/test_data_parallel.py]: done.

**Comment by rohitgr7** [comment]
All done. I just have one question: https://github.com/PyTorchLightning/pytorch-lightning/pull/6139#discussion_r580527338

File: tests/overrides/test_data_parallel.py (hunk 3 above)
Comment (kaushikb11) [reply tests/overrides/test_data_parallel.py:73]: You could use something like this https://github.com/kaushikb11/pytorch-lightning/blob/hooks%2Foutputs/tests/deprecated_api/__init__.py#L29

Comment (rohitgr7) [reply tests/overrides/test_data_parallel.py:73]: will do. thanks!

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/overrides/base.py:59]: Most probably because this was added and nobody remembered the other one :D

Although IMO it fits better in the training_loop than here

File: tests/helpers/utils.py [hunk 5]
```diff
@@ -111,3 +115,18 @@ def inner_f(queue, **kwargs):
...
+
+
+@contextmanager
+def no_warning_call(warning_type, match: Optional[str] = None):
+    with pytest.warns(None) as record:
+        yield
+
+        try:
+            w = record.pop(warning_type)
+            if not ((match and match in str(w.message)) or w):
```
Comment (carmocca) [inline tests/helpers/utils.py:127]: I think these parentheses are redundant. Also that there is `.text` but haven't run it

```suggestion
            if not (match and match in w.text or w):
```

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/overrides/base.py:59]: yeah, since this one is somewhat specific to distributed training, just wanted to confirm whether to remove it or not. @awaelchli

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/overrides/base.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/base.py:59]: hehe it just looks like I added it but it was a refactor :) It was there originally. 
This is a leftover artifact of how we called the training_step before in distributed and DP setting but now it should not be needed IF we already do it in the training loop. 

For DP I will check tomorrow how it behaves, not sure there because of the gather... Have to test.

Comment (rohitgr7) [reply pytorch_lightning/overrides/base.py:59]: hi @awaelchli, did you check this for DP?

Comment (awaelchli) [reply pytorch_lightning/overrides/base.py:59]: @rohitgr7  sorry for the delay. 
I tested and I get two different warnings for DP:

```
/home/adrian/.conda/envs/lightning/lib/python3.7/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: Your training_step returned None. Make sure that was your intention!
  warnings.warn(*args, **kwargs)


/home/adrian/.conda/envs/lightning/lib/python3.7/site-packages/pytorch_lightning/utilities/distributed.py:50: UserWarning: training_step returned None if it was on purpose, ignore this warning...
  warnings.warn(*args, **kwargs)

```
It should be fine to remove the one in the module completely (we can still have tests for the warning in the training loop).

Comment (rohitgr7) [reply pytorch_lightning/overrides/base.py:59]: cool! thanks.. will update.

**Comment by Borda** [comment]
@rohitgr7 is this still WIP? 🐰

File: pytorch_lightning/overrides/base.py [hunk 6]
```diff
@@ -54,19 +54,18 @@ def forward(self, *inputs, **kwargs):
...
         elif running_stage == RunningStage.TESTING:
             output = self.module.test_step(*inputs, **kwargs)
-            warn_if_output_is_none(output, "test_step")
 
         elif running_stage == RunningStage.EVALUATING:
             output = self.module.validation_step(*inputs, **kwargs)
-            warn_if_output_is_none(output, "validation_step")
 
         elif running_stage == RunningStage.PREDICTING:
             output = self.module.predict(*inputs, **kwargs)
```
Comment (tchaton) [inline pytorch_lightning/overrides/base.py:68]: We should warn during predict call as we expect the user to return the predictions.

Comment (rohitgr7) [reply pytorch_lightning/overrides/base.py:68]: added.

File: CHANGELOG.md [hunk 7]
```diff
@@ -15,6 +15,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `checkpoint` parameter to callback's `on_save_checkpoint` hook ([#6072](https://github.com/PyTorchLightning/pytorch-lightning/pull/6072))
 
 
+- Added no return warning to predict ([#6139](https://github.com/PyTorchLightning/pytorch-lightning/pull/6139))
```
Comment (awaelchli) [inline CHANGELOG.md:18]:
```suggestion
- Added warning when predict returns no predictions ([#6139](https://github.com/PyTorchLightning/pytorch-lightning/pull/6139))
```
This is misleading. You don't say "added no wheels to car to drive" xD

File: CHANGELOG.md [hunk 8]
```diff
@@ -26,6 +29,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed support for passing a bool value to `profiler` argument of Trainer ([#6164](https://github.com/PyTorchLightning/pytorch-lightning/pull/6164))
 
 
+- Removed no return warning from val/test step ([#6139](https://github.com/PyTorchLightning/pytorch-lightning/pull/6139))
```
Comment (awaelchli) [inline CHANGELOG.md:32]:
```suggestion
- Removed warning message for val/test step with no return value ([#6139](https://github.com/PyTorchLightning/pytorch-lightning/pull/6139))
```

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
lgtm

File: CHANGELOG.md (hunk 7 above)
Comment (rohitgr7) [reply CHANGELOG.md:18]: :laughing: .... thanks :+1:

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by carmocca** [comment]
@rohitgr7 can you check if this fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6339 and if not, also fix it in this PR?
